src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs changeEvent)
+        {
+            if (changeEvent.Kind == WorkspaceChangeKind.DocumentChanged
+                || changeEvent.Kind == WorkspaceChangeKind.DocumentAdded
+                || changeEvent.Kind == WorkspaceChangeKind.ProjectAdded)
+            {
+                var project = changeEvent.NewSolution.GetProject(changeEvent.ProjectId);
+                QueueForAnalysis(new[] { project });
+            }
         }
 
-        private async Task Dequeue()
+        private async Task Analyze(Project project, CancellationTokenSource workReadySource, CancellationToken token)
         {
-            var tasks = new List<Task<DiagnosticResult>>();
-            for (var i = 0; i < 50; i++)
+            try
             {
-                if (_openDocuments.IsEmpty)
+                // Only basic syntax check is available if file is miscellanous like orphan .cs file.
+                // Todo: Where this magic string should be moved?
+                if (project.Name == "MiscellaneousFiles")
                 {
-                    break;
+                    await AnalyzeSingleMiscFilesProject(project);
+                    return;
                 }
 
-                if (_openDocuments.TryDequeue(out var filePath))
+                var allAnalyzers = _providers
+                    .SelectMany(x => x.CodeDiagnosticAnalyzerProviders)
+                    .Concat(project.AnalyzerReferences.SelectMany(x => x.GetAnalyzersForAllLanguages()));
+
+                var compiled = await project.WithCompilationOptions(
+                    _rulesetsForProjects.BuildCompilationOptionsWithCurrentRules(project))
+                    .GetCompilationAsync(token);
+
+                ImmutableArray<Diagnostic> results = ImmutableArray<Diagnostic>.Empty;
+
+                if (allAnalyzers.ToImmutableArray().Any())
                 {
-                    tasks.Add(this.ProcessNextItem(filePath));
+                    results = await compiled
+                        .WithAnalyzers(allAnalyzers.ToImmutableArray()) // This cannot be invoked with empty analyzers list.
+                        .GetAllDiagnosticsAsync(token);
+                }
+                else
+                {
+                    results = compiled.GetDiagnostics();
                 }
-            }
 
-            if (!tasks.Any()) return;
+                _results[project.Id] = (project.Name, results);
 
-            var diagnosticResults = await Task.WhenAll(tasks.ToArray());
-            if (diagnosticResults.Any())
+                EmitDiagnostics(results);
+            }
+            finally
             {
-                var message = new DiagnosticMessage()
-                {
-                    Results = diagnosticResults
-                };
-
-                this._forwarder.Forward(message);
+                workReadySource.Cancel();
```

Comment:
All those tokens are actually for peding api requests not for compilation itself, api waits when it has response for projects it asks for (those which are pending at that point api call waits to get ready). However during that request same project can be requeued again but api call won't wait for it anymore. This is complexity from behavior how api is polled. It also has timeout mechanisms and logging for it to prevent complete hang and diagnose those issues.

If i remember correctly in version where analysis was aborted and new started again when files changed vscode atleast continously missed changes, basically that meaned that there was errors in invalid location and they were fixed after i made any action which made re-request current diagnostics from omnisharp by vscode. I think real place to fix this problem is vscode addon, but since it likely works much like other consumers of api (?) i am not eager to made those changes. LSP based diagnostics with events should simplify this once they are available.

If theres is file based api chat requests diagnostics from single file with current analyzers based on current best build it should work, however i didn't find any that give results with custom analyzers.